### PR TITLE
make WorkPackage.count_and_group_by private

### PR DESCRIPTION
**Pull request by tessi**
WorkPackage.count_and_group_by is a WorkPackage internal method
and it is currently only used as an internal method.
It allows SQL injection when the parameters come from user input, but
every known call serves safe constants as parameters to this method.

Anyways, we're better "safe" than sorry and make this method private.

[`* `#5590` Evaluate known security risks`](https://www.openproject.org/work_packages/5590)

### CODE REVIEW

File: app/models/work_package.rb [hunk 1]
```diff
@@ -875,45 +875,45 @@ def self.update_versions_from_hierarchy_change(project)
 
   # Extracted from the ReportsController.
   def self.by_type(project)
-    count_and_group_by(:project => project,
+    count_and_group_by :project => project,
                        :field => 'type_id',
-                       :joins => Type.table_name)
+                       :joins => Type.table_name
```
Comment (mfrister) [inline app/models/work_package.rb:880]: Do we need to remove all these braces? I'd say this is more of a taste thing and I'd rather not start changing taste things, especially if they're unrelated to the actual change.

Comment (tessi) [inline app/models/work_package.rb:880]: I don't care about the braces. It just happened while editing the code and I left it there.

Comment (mfrister) [inline app/models/work_package.rb:880]: Ok, then I'll just revert that part of the commit :)

Comment (myabc) [inline app/models/work_package.rb:880]: I think whitespace tweaks are fine :-) but I think it's generally good to aim for atomic commits. As such, it might be better to make whitespace changes in a separate commit (or even PR).

**Comment by mfrister** [comment]
Ok, as a compromise I split the changes up into two commits.

:+1:

**Comment by myabc** [comment]
:+1:

File: app/models/work_package.rb [hunk 2]
```diff
@@ -1020,13 +1020,15 @@ def close_duplicates
...
+  # based on specific criteria.
+  # DANGER: :field and :joins MUST never come from user input, because
+  # they are not SQL-escaped.
   #
   # Options
   # * project - Project to search in.
   # * field - String. Issue field to key off of in the grouping.
   # * joins - String. The table name to join against.
-  def self.count_and_group_by(options)
+  private_class_method def self.count_and_group_by(options)
```
Comment (myabc) [inline app/models/work_package.rb:1031]: I forgot you could specify private class methods in this way (I've always used this method with a symbol (i.e. `private_class_method: count_and_group_by`) :-)

Comment (tessi) [inline app/models/work_package.rb:1031]: This is actually a pretty new feature: https://www.ruby-lang.org/en/news/2013/12/25/ruby-2-1-0-is-released/

https://bugs.ruby-lang.org/issues/3753

considering this, it might be a bummer, if we want to support ruby 2.0..
